hassio/homeassistant.py

```diff
@@ -317,51 +332,90 @@ def in_progress(self):
             return ConfigResult(False, log)
         return ConfigResult(True, log)
 
-    async def check_api_state(self):
-        """Check if Home-Assistant up and running."""
-        url = f"{self.api_url}/api/"
-        header = {CONTENT_TYPE: CONTENT_TYPE_JSON}
+    # PyLint doesn't understand async gen
+    # pylint: disable=E1701
 
-        if self.api_password:
-            header.update({HEADER_HA_ACCESS: self.api_password})
+    async def ensure_access_token(self):
+        """Ensures there is an access token."""
+        if self.access_token is not None:
+            return
+
+        async with self.sys_websession_ssl.get(
+                f"{self.api_url}/auth/token",
+                timeout=30,
+                data={
+                    "grant_type": "refresh_token",
+                    "refresh_token": self.refresh_token
+                }
+        ) as resp:
+            if resp.status != 200:
+                raise HTTPUnauthorized()
+            tokens = await resp.json()
+            self.access_token = tokens['access_token']
+
+    @asynccontextmanager
+    async def make_request(self, method, path, json=None, content_type=None,
+                           data=None, timeout=30):
+        """Async context manager to make a request with right auth."""
+        url = f"{self.api_url}/{path}"
+        headers = {}
+
+        if content_type is not None:
+            headers[hdrs.CONTENT_TYPE] = content_type
+
+        if self.refresh_token:
+            await self.ensure_access_token()
+            headers[hdrs.AUTHORIZATION] = f'Bearer {self.access_token}'
+
+        elif self.api_password:
+            headers[HEADER_HA_ACCESS] = self.api_password
+
+        async with getattr(self.sys_websession_ssl, method)(
+                url,
+                timeout=timeout,
+                json=json,
+        ) as resp:
+            if resp.status != 401 or not self.refresh_token:
+                yield resp
+                return
+
+            # Access token expired
+            self.access_token = None
+            async with self.make_request(
+                    method, path, json, content_type, data, timeout,
+            ) as resp:
+                yield resp
```

Comment:
I think we should loop that and use the token refresh on start of that loop. If home assistant have a problem, we crash on ARM after some reqursive calls while the stack size is very small.